### PR TITLE
icon2html optimizations

### PR DESCRIPTION
**Pull request by Dejaku51**
## About The Pull Request
Ports https://github.com/tgstation/tgstation/pull/67429
ports https://github.com/tgstation/tgstation/pull/58607

## Why It's Good For The Game
Faster code is better

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

1.do something that calls icon2html
2. see that it runs faster than previously
## Changelog
:cl: Dejaku51, Kylerace, LemonInTheDark
refactor: Optimized icon2html code
/:cl:

### CODE REVIEW

**Review by itsmeow — APPROVED** [review]
Looks good.

File: code/modules/asset_cache/asset_cache_item.dm [hunk 1]
```diff
@@ -21,10 +25,20 @@
...
+///pass in a valid dmi file path string e.g. "icons/path/to/dmi_file.dmi" to make generating the hash less expensive
+/datum/asset_cache_item/New(name, file, file_hash, dmi_file_path)
 	if (!isfile(file))
 		file = fcopy_rsc(file)
-	hash = md5(file)
+
+	hash = file_hash
+
+	//the given file is directly from a dmi file and is thus in the rsc already, we know that its file_hash will be correct
+	if(!hash)
```
Comment (Archanial) [inline code/modules/asset_cache/asset_cache_item.dm:37]: doubled check

Comment (Archanial) [reply code/modules/asset_cache/asset_cache_item.dm:37]: actually whole logic is doubled here, looks like bad copypaste

**Review by Archanial — CHANGES_REQUESTED** [review]

File: code/modules/asset_cache/asset_cache_item.dm (hunk 1 above)
Comment (itsmeow) [reply code/modules/asset_cache/asset_cache_item.dm:37]: it's definitely not doubled wut

**Review by Archanial — APPROVED** [review]

**Pull request merged**